Model system

SRFU-NN opened the pull request:
Overhaul of the model systems.

* Fixes to avoid correlated noise in DataDependentNoise and SumNoise
* Creating and exporting ModelSystem's for benchmarks.
* Switched to local imports to avoid circular import
* NoiseModel._noise_distribution is now a method, to allow changes of self._rng to affect self._noise_distribution automatically

SRFU-NN committed:
feat: Populated model_sysmte.__init__.py

Also switched to local imports to avoid problems with circular imports.

SRFU-NN committed:
feat: Model system has noise size

SRFU-NN committed:
fix: Branin-Hoo model system works

SRFU-NN committed:
doc: Use get_random_generator when available

SRFU-NN committed:
style: NoiseModel._noise_distribution is property

By having it as a property method, it can be dynamically changed when
self._rng changes.

SRFU-NN committed:
test: Updated tests

SRFU-NN committed:
fix: Corrected NoiseModel._sample_noise

SRFU-NN committed:
feat: Allow noise models to overwrite rng

DataDependentNoise and SumNoise has underlying noise models. If these
have independent rng's that are seeded with the same seed, the noise
does not behave as wanted. Default behavior is now that the overlying
nosie model overwrites the rng's of the underlying noise models with
its own.

SRFU-NN committed:
doc: Jupyter notebook about ModelSystem

SRFU-NN committed:
style: Black formatting

SRFU-NN committed:
feat: ModelSytem for hart3, hart6, poly2, peaks

SRFU-NN committed:
style: Removed unused import

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into model_system

SRFU-NN commented on ProcessOptimizer/model_systems/__init__.py:
```diff
@@ -0,0 +1,31 @@
+    parse_noise_model,
+    noise_model_factory,
+)
+from .benchmarks import hart3_model_system as hart3
```
This isn't very nice - I am importing the model system, and renaming it to have the same name as the benchmark function has in `benchmarks.py`. This could cause confusion, as

`from model_systems import hart3`

and

`from model_systems.benchmarks import hart3`

returns very different objects, but both are named `hart3`, and it is not super transparent why.

I want the user to automatically get the model system, and not even notice that there is a benchmark function, but also keep the benchmark function to be imported exactly as is, for backward compatibility. Please comment if my approach is acceptable.

We could also simply

`from .benchmarks import hart3_model_system`

, but I would like the "thing" the user gets to be a model system that is named after the benchmark function.

dk-teknologisk-mon replied:
Hmm, I'm not convinced that we have a plethora of users that make extensive use of the benchmark _functions_ themselves yet, mainly because most of them are quite recent additions (peaks + hartmann functions). Couldn't we just rename the functions inside `benchmarks.py` to `hart3_function`, `peaks_function` etc. since this would make the distinction explicit and the code that builds the model systems more readable? Then we can keep the short names like `hart3` reserved for the thing that we intend for the users to interact with, namely the model systems.

SRFU-NN replied:
I would prefer that solution, but didn't want to break backward compatibility, in case anyone WAS using the benchmarks functions.

I don't think I will get time to finish this before my holiday, so you have three weeks to mull it over :-)

dk-teknologisk-mon replied:
So the consequence would be to expand the names of the soon-to-be entirely internal function definitions inside benchmarks.py and shorten the names of the actual model systems down.

SRFU-NN replied:
If we are to change that, we could also consider splitting them into separate files, like the one named `gold_map.py`. It is easier to find them that way. It doesn't change anything for the user, since they are still imported into `__init__.py`, so it would still be `from ProcessOptimizer.model_systems import XXXX`.

We get quite a lot of files in that folder, but I don't see that as a big issue.

dk-teknologisk-mon replied:
I don't mind that implementation either, so feel free to go for that :)

SRFU-NN replied:
I will, then, but I won't have time before my holiday. I will get on it in August.

SRFU-NN replied:
@dk-teknologisk-mon : Done, would you give it a new review?

SRFU-NN committed:
fix: Re-calculated notebook to get fixed output

SRFU-NN committed:
feat: Added gold map as a model system

SRFU-NN committed:
doc: Changelog

SRFU-NN committed:
Merge remote-tracking branch 'origin/develop' into model_system

SRFU-NN commented:
I am having problems with circular imports in utils (I thought I had fixed it, but it is back).

Looking at it, Space is imported into utils in order to make `normalize_dimensions()` and `cook_estimator()` work.

Do these really belong in utils? `nomralize_dimensions()` seems like it is intimately about Dimensions, so doesn't it belong in `..space``? 

And `cook_estimator` seems to be about regressors and kernels, so doesn't it belong in `..learning`?

That way, we can limit the amount of import from different parts of the package into `utils`, which is going to make it easier to use it.

We should probalby also move `has_gradients()` too `..learning`, since it is about estimators.

SRFU-NN committed:
refactor: moved functions out of utils

To avoid circular import errors,` cook_estimator`, `has_gradients`,
and `use_named_args` are moved to `learning`, and `normalize_dimensions`
is moved to `space`.

SRFU-NN committed:
chore: Removed replaced functions

SRFU-NN committed:
doc: Updated changelog

SRFU-NN commented:
I have fixed the circular import, but it involved moving `cook_estimator`, `has_gradients`, `use_named_args`, and `normalize_dimensions` out of utils, breaking any direct imports of them. The first three are moved to `learning`, which makes the tests less readable, as I have to exclude them from being predictors.

We could consider not exporting them into the __init__.py file, so `ProcessOptimizer.learning` only contains predictors (?). This  would make the import of them more cumbersome, but if no users use them, I think that could be worthwile.

SRFU-NN committed:
doc: Spelling.

SRFU-NN committed:
refactor: Branin-Hoo changed to a `ModelSystem`

SRFU-NN committed:
test: Fixed errors in testing

SRFU-NN committed:
feat: None accepted by `parse_noise_model`

None is intepreted as zero noise.

SRFU-NN committed:
test: Fixed branin used in `test_common.py`

Also introduced `branin_no_noise`

SRFU-NN committed:
test: Fixed imports of branin into tests

SRFU-NN committed:
refactor: Using the builtin no noise Branin-Hoo

Insted of modifying the deafult Branin-Hoo `ModelSystem` to not have
noise, import the one that doesn't have any noise and us it in tests.

SRFU-NN committed:
chore: Added None as option for noise_model

In type hints

SRFU-NN committed:
style: hart3 calling simplified

SRFU-NN committed:
style: Removed unused import

SRFU-NN committed:
refactor:  Streamlined  `hart6` `ModelSystem`

SRFU-NN committed:
chore: Removed old `hart6`

SRFU-NN committed:
fix: removed import of old `hart6`

SRFU-NN committed:
refactor: Moved poly2 to separate file

SRFU-NN committed:
feat: NoiseModel accepting "none" as zero noise

SRFU-NN committed:
refactor:  `peaks ` `ModelSystem` in separate file

SRFU-NN committed:
style: Removed unused imports

SRFU-NN committed:
doc: Updated changelog

dk-teknologisk-mon commented on ProcessOptimizer/utils/utils.py:
```diff
@@ -813,8 +456,16 @@ def y_coverage(res, return_plot=False, random_state=None, horizontal=False):
+            [
+                max_x,
+            ]
+        )
```
I think this particular change reduces readability when the lines are this simple.

SRFU-NN replied:
Agreed, will fix

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/__init__.py:
```diff
@@ -0,0 +1,38 @@
+from .branin_hoo import branin, branin_no_noise
```
Why is this particular function placed separately from the remaining model systems? I think it'd be prettier if they were all collected together below (line 12 and onwards).

SRFU-NN replied:
Agreed. note that this breaks the isort order, but I think the added readability here makes up for it. Will fix

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/branin_hoo.py:
```diff
@@ -1,20 +1,61 @@
+    branin_score,
+    [Real(-5, 10, name="x1"), Real(0, 15, name="x2")],
+    noise_model={"model_type": "constant", "noise_size": 0.02},
+    true_min=0.397887,
```
```suggestion
    noise_model={"model_type": "constant", "noise_size": 0.02},
    true_max=308.13,
    true_min=0.397887,
```

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/branin_hoo.py:
```diff
@@ -1,20 +1,61 @@
+    branin_score,
+    [Real(-5, 10, name="x1"), Real(0, 15, name="x2")],
+    noise_model=None,
+    true_min=0.397887,
```
```suggestion
    noise_model=None,
    true_max=308.13,
    true_min=0.397887,
```

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/branin_hoo.py:
```diff
@@ -1,20 +1,61 @@
+branin = ModelSystem(
+    branin_score,
+    [Real(-5, 10, name="x1"), Real(0, 15, name="x2")],
+    noise_model={"model_type": "constant", "noise_size": 0.02},
```
This noise size is _very_ low compared to the scale of the system (about 300 in absolute units). I assume it has this value because noise greater than this makes the local minima indistinguishable in practice, or is there some other reason for this value?

SRFU-NN replied:
We should probably go through all of the noise levels, I don't think I have given them enough thought.

And this is almost certainly too small.

SRFU-NN replied:
Made proportional to the signal, at the 10% level. This might seem high compared to the 1% default for constant noise, but the 1% is relative to the entire span, which mostly means the worst value, while the 1% is relative to the signal at the sampling point, which is much lower.

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/hart3.py:
```diff
@@ -0,0 +1,57 @@
+hart3 = ModelSystem(
+    hart3_score,
+    [(0.0, 1.0), (0.0, 1.0), (0.0, 1.0)],
+    noise_model="constant",
```
```suggestion
    noise_model={"model_type": "constant", "noise_size": 0.03863},
```
Assigns a default noise of 1 % of the system scale, otherwise this would default to noise_size=1, which is enormous.

SRFU-NN replied:
1% has been set as default noise:
```
from ProcessOptimizer import model_systems
model_systems.hart3.noise_size
```
yeilds 0.03863

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/hart6.py:
```diff
@@ -0,0 +1,68 @@
+hart6 = ModelSystem(
+    hart6_score,
+    [(0.0, 1.0) for _ in range(6)],
+    noise_model="constant",
```
```suggestion
    noise_model={"model_type": "constant", "noise_size": 0.033224},
```
Assigns a default noise of 1 % of the system scale, otherwise this would default to noise_size=1, which is enormous.

SRFU-NN replied:
1% has been set as default noise:
```
model_systems.hart6.noise_size
```
returns 0.033224000000000004

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/peaks.py:
```diff
@@ -0,0 +1,52 @@
+peaks = ModelSystem(
+    peaks_score,
+    [(-3.0, 3.0), (-3.0, 3.0)],
+    noise_model="constant",
```
```suggestion
    noise_model={"model_type": "constant", "noise_size": 0.1466},
```
Assigns a default noise of 1 % of the system scale.

SRFU-NN replied:
1% has been set as default noise:
```
model_systems.peaks.noise_size
```
yields 0.146571

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/poly2.py:
```diff
@@ -0,0 +1,52 @@
+poly2 = ModelSystem(
+    poly2_score,
+    [(-1.0, 1.0), (-1.0, 1.0)],
+    noise_model="constant",
```
```suggestion
    noise_model={"model_type": "constant", "noise_size": 0.007812},
```
Assigns a default noise of 1 % of the system scale.

SRFU-NN replied:
1% has been set as default noise:
```
model_systems.poly2.noise_size
```
returns 0.007812000000000001

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/poly2.py:
```diff
@@ -0,0 +1,52 @@
+poly2_no_noise = ModelSystem(
+    poly2_score,
+    [(-1.0, 1.0), (-1.0, 1.0)],
+    noise_model="none",
```
```suggestion
    noise_model=None,
```

dk-teknologisk-mon commented on ProcessOptimizer/model_systems/gold_map.py:
```diff
@@ -0,0 +1,27 @@
+gold_map = ModelSystem(
+    score,
+    space=[(0.0, 15.0), (0.0, 15.0)],
+    noise_model="zero",
```
```suggestion
    noise_model=None,
```
We don't have to use `None`, I would just like us to consistently use any of `None, "none" or "zero"` in across all the model system versions where we don't add noise. I have suggested `None` across the board.

SRFU-NN replied:
Yeah, I was getting cute with the different ways to do that. I'll add tests for each of them instead.

dk-teknologisk-mon commented on ProcessOptimizer/optimizer/optimizer.py:
```diff
@@ -1180,33 +1120,48 @@ def NSGAII(self, MU=40):
+            and self.models[-1].noise != "gaussian"
+        ):
+            raise ValueError(
+                "expected noise to be 'gaussian', got %s" % self.models[-1].noise
```
```suggestion
                "Expected noise to be 'gaussian', got %s" % self.models[-1].noise
```

dk-teknologisk-mon commented on ProcessOptimizer/space/space.py:
```diff
@@ -111,8 +111,9 @@ def check_dimension(dimension, transform=None):
     if len(dimension) == 3:
-        if (any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and 
-                dimension[2] in ["uniform", "log-uniform"]):
+        if any([isinstance(dim, (float, int)) for dim in dimension[:2]]) and dimension[
+            2
+        ] in ["uniform", "log-uniform"]:
```
This change is really weird. I suggest keeping the original code.

SRFU-NN replied:
Done, moving `and` to the new line to make it more prominant.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_individual_model_systems.py:
```diff
@@ -0,0 +1,64 @@
+class TestModelSystem(unittest.TestCase):
+    # Make a test for the branin ModelSystem
+    def test_branin(self):
+        assert branin.noise_size == 0.02
```
If you change the default value in the ModelSystem, it should also be changed here.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_parallel_cl.py:
```diff
@@ -8,23 +8,31 @@
 from scipy.spatial.distance import pdist
 import pytest
 
+# Introducing branin function as a test function from hte Branin no noise ModelSystem
```
```suggestion
# Introducing branin function as a test function from the Branin no noise ModelSystem
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_plot.py:
```diff
@@ -114,33 +114,38 @@ def test_plot_objective():
+        "uncertain_color": [1, 1, 1],
+        "colormap": "viridis_r",
+        "normalize_uncertainty": (
+            lambda x, global_min, global_max: (x - global_min)
+            / (global_max - global_min)
```
The linebreak here is a bit weird to read.

SRFU-NN replied:
Agree that this is weird, but Black insists on this. Will fix in codespace once all else has been agreed upon.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_utils.py:
```diff
@@ -12,17 +12,18 @@
 from ProcessOptimizer import Space
 from ProcessOptimizer.utils import point_asdict
 from ProcessOptimizer.utils import point_aslist
 from ProcessOptimizer.utils import dimensions_aslist
```
```suggestion
from ProcessOptimizer.utils import (
    point_asdict,
    point_aslist,
    dimensions_aslist,
)
```

SRFU-NN replied:
Fixed, except I have kept it on one line, as it doesn't get too long.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_utils.py:
```diff
@@ -12,17 +12,18 @@
-from ProcessOptimizer.utils import normalize_dimensions
-from ProcessOptimizer.utils import use_named_args
-from ProcessOptimizer.space import Real, Integer, Categorical
+from ProcessOptimizer.space import normalize_dimensions, Real, Integer, Categorical
```
```suggestion
from ProcessOptimizer.space import (
    normalize_dimensions,
    Real,
    Integer,
    Categorical,
)
```

SRFU-NN replied:
Black suggests having them on the same line, as the line doesn't get too long. I think I prefer keeping them on one line, it is not like a list of arguments where you have to keep track of the order.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_utils.py:
```diff
@@ -122,45 +131,39 @@ def test_expected_minimum_max():
+    )
+    X = [[-2, "A"], [-1, "A"], [0, "A"], [1, "A"], [2, "A"]]
+    y = [2, 1, 0, 1, 2]
+    result = opt.tell(X, y)
```
```suggestion
    x = [[-2, "A"], [-1, "A"], [0, "A"], [1, "A"], [2, "A"]]
    y = [2, 1, 0, 1, 2]
    result = opt.tell(x, y)
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_utils.py:
```diff
@@ -122,45 +131,39 @@ def test_expected_minimum_max():
+    )
+    X = [[-2, "A"], [-1, "A"], [0, "A"], [1, "A"], [2, "A"]]
+    y = [2, 1, 0, 1, 2]
+    result = opt.tell(X, y)
```
```suggestion
    x = [[-2, "A"], [-1, "A"], [0, "A"], [1, "A"], [2, "A"]]
    y = [2, 1, 0, 1, 2]
    result = opt.tell(x, y)
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -10,30 +10,46 @@
-ACQ_FUNCS_MIXED = ["EI"] #, "EIps" removed
-ESTIMATOR_STRINGS = ["GP", "RF", "ET", "GBRT", "DUMMY",
-                     "gp", "rf", "et", "gbrt", "dummy"]
+# Introducing branin function as a test function from hte Branin no noise ModelSystem
```
```suggestion
# Introducing branin function as a test function from the Branin no noise ModelSystem
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -131,11 +164,16 @@ def test_dimension_checking_2D():
     opt = Optimizer([(low, high), (low, high)])
     # within bounds but one dimension too little
     with pytest.raises(ValueError) as e:
-        opt.tell([low+1, ], 2.)
+        opt.tell(
+            [
+                low + 1,
+            ],
+            2.0,
+        )
```
This change reduces readability. I suggest keeping the original.

SRFU-NN replied:
Yeah, that is weird. Will change back

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -146,21 +184,32 @@ def test_dimension_checking_2D_multiple_points():
     opt = Optimizer([(low, high), (low, high)])
     # within bounds but one dimension too little
     with pytest.raises(ValueError) as e:
-        opt.tell([[low+1, ], [low+1, low+2], [low+1, low+3]], 2.)
+        opt.tell(
+            [
+                [
+                    low + 1,
+                ],
+                [low + 1, low + 2],
+                [low + 1, low + 3],
+            ],
+            2.0,
+        )
```
This change reduces readability. I suggest keeping the original.

SRFU-NN replied:
Agreed

SRFU-NN replied:
It is the commas after the last element that triggers Black to make line changes. `[low+1, ]` has been changed to `[low+1]`, which should return the same list.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -146,21 +184,32 @@ def test_dimension_checking_2D_multiple_points():
-    opt = Optimizer([(-2.0, 2.0)], base_estimator, n_initial_points=2,
-                    acq_optimizer="sampling")
-    result = opt.tell([1.5], 2.)
+    opt = Optimizer(
+        [(-2.0, 2.0)], base_estimator, n_initial_points=2, acq_optimizer="sampling"
+    )
+    result = opt.tell([1.5], 2.0)
```
```suggestion
    opt = Optimizer(
        [(-2.0, 2.0)],
        base_estimator,
        n_initial_points=2,
        acq_optimizer="sampling",
    )
    result = opt.tell([1.5], 2.0)
```

SRFU-NN replied:
There's quite a lot of these, I have changed them all (I think).

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -274,31 +336,31 @@ def test_exhaust_initial_calls(base_estimator):
-    opt.run(func=lambda x: x[0]**2, n_iter=3)
+    opt = Optimizer(
+        [(-2.0, 2.0)], base_estimator=base_estimator, n_initial_points=2, acq_func="EI"
+    )
```
```suggestion
    opt = Optimizer(
        [(-2.0, 2.0)],
        base_estimator=base_estimator,
        n_initial_points=2,
        acq_func="EI",
    )
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_optimizer.py:
```diff
@@ -336,60 +398,51 @@ def test_iterating_ask_tell_lhs():
 
 @pytest.mark.slow_test
 def test_add_remove_modelled_noise():
-    '''
+    """
     Tests whether the addition of white noise leads to predictions closer to
-    known true values of experimental noise (iid gaussian noise)'''
+    known true values of experimental noise (iid gaussian noise)"""
+
     # Define objective function
     def flat_score(x):
         return 42
-    
+
     # Set noise and model system
     noise_size = 0.45
     flat_space = [(-1.0, 1.0)]
     flat_noise = {
         "model_type": "constant",
         "noise_size": noise_size,
-        }
+    }
     # Build ModelSystem object
     model = ModelSystem(
         score=flat_score,
         space=flat_space,
         noise_model=flat_noise,
-        )
+    )
     # Instantiate Optimizer
-    opt = Optimizer(
-        flat_space,
-        "GP",
-        lhs=False,
-        n_initial_points=1,
-        random_state=42
-        )
-    #Make 20 dispersed points on X
+    opt = Optimizer(flat_space, "GP", lhs=False, n_initial_points=1, random_state=42)
+    # Make 20 dispersed points on X
```
I suggest sticking with original formatting for consistency across the tests.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_common.py:
```diff
@@ -150,12 +158,18 @@ def test_fixed_random_states(minimizer):
     space = [(-5.0, 10.0), (0.0, 15.0)]
-    result1 = minimizer(branin, space, n_calls=n_calls,
-                        n_random_starts=n_random_starts, random_state=1)
+    result1 = minimizer(
+        branin, space, n_calls=n_calls, n_random_starts=n_random_starts, random_state=1
+    )
```
```suggestion
    result1 = minimizer(
        branin,
        space,
        n_calls=n_calls,
        n_random_starts=n_random_starts,
        random_state=1,
    )
```

SRFU-NN replied:
The original is within the 88 characters a line is allowed to be, which is why it isn't broken. But honestly, it is easier to read your way. Will change.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_common.py:
```diff
@@ -170,24 +184,31 @@ def test_minimizer_with_space(minimizer):
     space = Space([(-5.0, 10.0), (0.0, 15.0)])
-    space_result = minimizer(branin, space, n_calls=n_calls,
-                             n_random_starts=n_random_starts, random_state=1)
+    space_result = minimizer(
+        branin, space, n_calls=n_calls, n_random_starts=n_random_starts, random_state=1
+    )
```
```suggestion
    space_result = minimizer(
        branin,
        space,
        n_calls=n_calls,
        n_random_starts=n_random_starts,
        random_state=1,
    )
```

SRFU-NN replied:
See previous comment

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_common.py:
```diff
@@ -376,53 +408,60 @@ def test_consistent_x_iter_dimensions(minimizer):
-                    n_random_starts=0)
+    res = minimizer(
+        bench1, dimensions=[(0, 1)], x0=[[0], [1]], n_calls=3, n_random_starts=0
+    )
```
```suggestion
    res = minimizer(
        bench1,
        dimensions=[(0, 1)],
        x0=[[0], [1]],
        n_calls=3,
        n_random_starts=0,
    )
```

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_common.py:
```diff
@@ -376,53 +408,60 @@ def test_consistent_x_iter_dimensions(minimizer):
     with pytest.raises(RuntimeError):
-        minimizer(bench1, dimensions=[(0, 1)],
-                  x0=[[0, 1]], n_calls=3, n_random_starts=0)
+        minimizer(
+            bench1, dimensions=[(0, 1)], x0=[[0, 1]], n_calls=3, n_random_starts=0
+        )
```
```suggestion
        minimizer(
            bench1,
            dimensions=[(0, 1)],
            x0=[[0, 1]],
            n_calls=3,
            n_random_starts=0,
        )
```

SRFU-NN replied:
I'll go through the test file and change it where I find it. My gut feeling is that five or more arguments are easier to read when they are on separate lines.

dk-teknologisk-mon commented on ProcessOptimizer/tests/test_common.py:
```diff
@@ -376,53 +408,60 @@ def test_consistent_x_iter_dimensions(minimizer):
     with pytest.raises(RuntimeError):
-        minimizer(bench1, dimensions=[(0, 1)],
-                  x0=[0, 1], n_calls=3, n_random_starts=0)
+        minimizer(bench1, dimensions=[(0, 1)], x0=[0, 1], n_calls=3, n_random_starts=0)
```
```suggestion
        minimizer(
            bench1,
            dimensions=[(0, 1)],
            x0=[0, 1],
            n_calls=3,
            n_random_starts=0,
        )
```

SRFU-NN committed:
style: Code review suggestions implemented

SRFU-NN committed:
feat: Default noise size for `ModelSystem` 1%

If o nosie size is given for a `ModelSystem` with constant noise, the
noise size is set to 1% of the span of the score function, interpreted
`true_max-true_min`

SRFU-NN committed:
style: Code review suggestions implemented

SRFU-NN commented:
![image](https://github.com/novonordisk-research/ProcessOptimizer/assets/104423389/b602682e-eda1-4ba6-9b21-7e7bac0642fa)
Noise levels, and equality of two ways to write a list.

dk-teknologisk-mon reviewed (approved):
Right, everything looks really good to me now so I will go ahead and merge this. Really great work! :)

SRFU-NN committed:
style: Circumventing Black

SRFU-NN commented:
> Right, everything looks really good to me now so I will go ahead and merge this. Really great work! :)

Thank you, and great work on the review, that was a large one!
